### PR TITLE
Oliinyk / PR2198 Improvement/Fix

### LINKED ISSUE #2113: [Personal cabinet] The service provider cannot send the message.
**Environment:** Windows 10, 64-bit, Google Chrome 107.0.5304.107.
**Reproducible:** always.

**Preconditions**
1.Go to https://oos.dmytrominochkin.cloud/#/
2.Log in as service provider.
3.SP has statements in his personal cabinet.

**Steps to reproduce**
1.Go to the "Заявки" tab of personal cabinet (by clicking the button in the top right corner).
2.Pay attention to the missing letter icons on the "Заявки" page.

**Actual result**
In the service provider's cabinet on the "Заявки" page, there is no possibility to send messages to contact users who have submitted applications for workshops.
Снимок экрана 2023-06-06 132734

**Expected result**
In the service provider's cabinet  on the "Заявки" page, at the end of each submitted application, there is a letter icon for sending messages to users who have submitted applications for workshops.
Снимок экрана 2023-06-07 121352

**Labels to be added**
"Bug", Priority ("medium"), Severity ("medium"), Type ("UI, "Functional").

### PR DESCRIPTION
**Pull request by doliinyk**
#2113 

Fixed issue when provider is trying to start chat with parent

### CODE REVIEW

File: src/app/shell/personal-cabinet/provider/provider-applications/provider-applications.component.ts [hunk 1]
```diff
@@ -169,6 +170,8 @@ export class ProviderApplicationsComponent extends CabinetDataComponent implemen
   }
 
   onSendMessage(application: Application): void {
+    this.store.dispatch(new SetParent(application.parent));
```
Comment (doliinyk) [inline src/app/shell/personal-cabinet/provider/provider-applications/provider-applications.component.ts]: Chat is working now, but have a doubt about such solution.
It is needed to have parent object (id, name) to start chat. So here it is dispatching to store using parent **in application**.

Maybe there is another approach or it is better to fetch parent from backend?

File: src/app/shell/personal-cabinet/shared-cabinet/messages/messages.component.ts [hunk 2]
```diff
@@ -42,6 +38,13 @@ export class MessagesComponent extends CabinetDataComponent {
   readonly WorkshopDeclination = WorkshopDeclination;
   readonly noMessagesTitle = NoResultsTitle.noMessages;
 
+  @Select(ProviderState.truncated)
+  workshops$: Observable<TruncatedItem[]>;
```
Comment (humanmantis) [inline src/app/shell/personal-cabinet/shared-cabinet/messages/messages.component.ts]: add access modifiers

File: src/app/shell/personal-cabinet/shared-cabinet/messages/chat/chat.component.ts [hunk 3]
```diff
@@ -81,60 +81,100 @@ export class ChatComponent implements OnInit, AfterViewInit, OnDestroy {
...
+    if (isUserDown) {
+      this.scrollDown();
     }
   }
 
-  public onBack(): void {
-    this.location.back();
+  private scrollDown(): void {
+    //TODO: Replace setTimeout
+    setTimeout(() => {
```
Comment (humanmantis) [inline src/app/shell/personal-cabinet/shared-cabinet/messages/chat/chat.component.ts]: use `asyncScheduler` instead

File: src/app/shell/personal-cabinet/shared-cabinet/messages/chat/chat.component.ts [hunk 4]
```diff
@@ -81,60 +81,100 @@ export class ChatComponent implements OnInit, AfterViewInit, OnDestroy {
...
+
+  private setChatSubscriptions(): void {
+    this.chatRoom$.pipe(filter(Boolean), takeUntil(this.destroy$)).subscribe((chatRoom: ChatRoom) => {
+      this.chatRoom = chatRoom;
+      this.store.dispatch(new GetChatRoomMessagesById(this.userRole, this.chatRoom.id, this.messagesParameters));
+      this.getChatMembersNames();
+
+      this.chatEl.nativeElement.onscroll = () => {
+        if (this.chatEl.nativeElement.scrollTop < 10) {
+          this.chatEl.nativeElement.scrollTop = 10;
```
Comment (humanmantis) [inline src/app/shell/personal-cabinet/shared-cabinet/messages/chat/chat.component.ts]: add readonly variable for `10`

File: src/app/shared/services/chat/chat.service.ts [hunk 5]
```diff
@@ -1,51 +1,61 @@
...
   getChatRooms(parameters: ChatRoomsParameters): Observable<SearchResponse<ChatRoom[]>> {
-    let params = this.setFilterParams(parameters);
+    const params = this.setChatRoomParams(parameters);
 
     return this.http.get<SearchResponse<ChatRoom[]>>(`/api/v1/ChatWorkshop/${parameters.role}/chatrooms`, { params });
   }
 
-  getChatRoomsMessages(chatRoomId: string, role: Role, parameters: MessagesParameters): Observable<IncomingMessage[]> {
-    let params = new HttpParams().set('Size', parameters.size.toString()).set('From', parameters.from.toString());
+  getChatRoomsForProviderByParentId(parentId: string): Observable<ChatRoom[]> {
```
Comment (humanmantis) [inline src/app/shared/services/chat/chat.service.ts]: add access modifiers for methods

File: src/app/shared/services/chat/chat.service.ts [hunk 6]
```diff
@@ -1,51 +1,61 @@
...
+    return this.http.get<ChatRoom[]>(`/api/v1/ChatWorkshop/provider/chatroomsforparent/${parentId}`);
```
Comment (humanmantis) [inline src/app/shared/services/chat/chat.service.ts]: `/api/v1/ChatWorkshop` this part can be a variable e.g. `baseApiUrl` or something similar

**Review by humanmantis — CHANGES_REQUESTED** [review]

File: src/app/shell/personal-cabinet/shared-cabinet/messages/chat/chat.component.ts [hunk 7]
```diff
@@ -81,60 +81,100 @@ export class ChatComponent implements OnInit, AfterViewInit, OnDestroy {
...
+    this.destroy$.unsubscribe();
+  }
+
+  private setChatSubscriptions(): void {
+    this.chatRoom$.pipe(filter(Boolean), takeUntil(this.destroy$)).subscribe((chatRoom: ChatRoom) => {
+      this.chatRoom = chatRoom;
+      this.store.dispatch(new GetChatRoomMessagesById(this.userRole, this.chatRoom.id, this.messagesParameters));
+      this.getChatMembersNames();
+
+      this.chatEl.nativeElement.onscroll = () => {
```
Comment (humanmantis) [inline src/app/shell/personal-cabinet/shared-cabinet/messages/chat/chat.component.ts]: use `addEventListener`, so we can call `removeEventListener` on destroy
another approach is Angular `Renderer2 tool

**Review by humanmantis — CHANGES_REQUESTED** [review]

**Review by humanmantis — APPROVED** [review]

**Pull request merged**